crates/nu-protocol/src/errors/shell_error.rs

```diff
@@ -108,6 +107,34 @@ pub enum ShellError {
         span: Span,
     },
 
+    /// A value's type did not match the expected type.
+    ///
+    /// ## Resolution
+    ///
+    /// Convert the value to the correct type or provide a value of the correct type.
+    #[error("Type mismatch")]
+    #[diagnostic(code(nu::shell::type_mismatch))]
+    RuntimeTypeMismatch {
+        expected: Type,
+        actual: Type,
+        #[label = "expected {expected}, but got {actual}"]
+        span: Span,
+    },
+
+    /// A value had the correct type but is otherwise invalid.
+    ///
+    /// ## Resolution
+    ///
+    /// Ensure the value meets the criteria in the error message.
+    #[error("Invalid value")]
+    #[diagnostic(code(nu::shell::invalid_value))]
+    InvalidValue {
```

Comment:
Here there is a confusion risk with the existing `IncorrectValue` that has freeform message but two spans.

What would be the rules people should follow to choose between the two.

Reply:
No rule/guideline is necessary. A PR right after this will remove `IncorrectValue`.

Reply:
Sounds good, `InvalidValue` sounds a bit better and if we nudge folks towards the "expected" / "got" pattern the better.